File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        l,r = 0,len(nums)-1

        while l<r:
            mid = (l+r) //2

            if nums[mid] < nums[r]:
                r = mid
            else :
                l = mid+1

        min_idx = l
        if target <= nums[len(nums)-1]:
            l = min_idx
            r = len(nums)-1
        else:
            l = 0
            r = min_idx
        
        while l<=r:
            mid = (l+r) //2

            if target < nums[mid]:
                r=mid-1
            elif target > nums[mid]:
                l=mid+1
            else:
                return mid
            
            

        return -1

File: Data Structures & Algorithms/search-2d-matrix/submission-1.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        
        for (int i= matrix.length-1;i>=0;i--){
            if (matrix[i][0]==target){
                return true;
            }

            if (matrix[i][0]<target){
                int left = 0;
                int right = matrix[i].length-1;

                while(left<=right){
                    int mid = (left+right)/2;

                    if (matrix[i][mid]>target){
                        right = mid-1;
                    }else if (matrix[i][mid]<target){
                        left = mid+1;
                    }else{
                        return true;
                    }
                }
            }
        }

        return false;
    }
}
